label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+          this.labelApplicationRuleStore.getObject(requestContext, ruleId);
+      if (optionalLabelApplicationRule.isEmpty()) {
+        responseObserver.onError(new StatusRuntimeException(Status.NOT_FOUND));
+        return;
+      }
+      responseObserver.onNext(
+          GetLabelApplicationRuleResponse.newBuilder()
+              .setLabelApplicationRule(optionalLabelApplicationRule.get())
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getLabelApplicationRules(
+      GetLabelApplicationRulesRequest request,
+      StreamObserver<GetLabelApplicationRulesResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      this.requestValidator.validateOrThrow(requestContext, request);
+      List<LabelApplicationRule> labelApplicationRules =
+          this.labelApplicationRuleStore.getAllObjects(requestContext);
+      responseObserver.onNext(
+          GetLabelApplicationRulesResponse.newBuilder()
+              .addAllLabelApplicationRules(labelApplicationRules)
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void updateLabelApplicationRule(
+      UpdateLabelApplicationRuleRequest request,
+      StreamObserver<UpdateLabelApplicationRuleResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      this.requestValidator.validateOrThrow(requestContext, request);
+      String ruleId = request.getId();
+      LabelApplicationRule labelApplicationRule =
+          LabelApplicationRule.newBuilder().setId(ruleId).setData(request.getData()).build();
+      if (this.labelApplicationRuleStore.getObject(requestContext, ruleId).isEmpty()) {
+        responseObserver.onError(new StatusRuntimeException(Status.NOT_FOUND));
+        return;
+      }
+      LabelApplicationRule updateLabelApplicationRule =
+          this.labelApplicationRuleStore.upsertObject(requestContext, labelApplicationRule);
+      responseObserver.onNext(
+          UpdateLabelApplicationRuleResponse.newBuilder()
+              .setLabelApplicationRule(updateLabelApplicationRule)
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void deleteLabelApplicationRule(
+      DeleteLabelApplicationRuleRequest request,
+      StreamObserver<DeleteLabelApplicationRuleResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      this.requestValidator.validateOrThrow(requestContext, request);
+      String ruleId = request.getId();
+      if (this.labelApplicationRuleStore.getObject(requestContext, ruleId).isEmpty()) {
```

Comment:
this check already happens lower down, doing it here is a redundant fetch.

Reply:
I don't think its a redundant fetch. Can you point me to the line where this check happens for the second time?

Reply:
https://github.com/hypertrace/config-service/blob/318a1e3fd88bbd3fa4cc832d9d4333b6679e3e54/config-service-impl/src/main/java/org/hypertrace/config/service/ConfigServiceGrpcImpl.java#L110-L116

Reply:
(by lower down I meant lower in the stack, not in this file 😉 )

Reply:
For some reason I am encountering a NullPointerException when trying to delete a non-existent config.

Reply:
Where's the NPE? What's null?

Reply:
Discussed offline. This functionality should be added to the mock config service.